src/ray/raylet/node_manager.cc

```diff
@@ -341,22 +342,41 @@ void NodeManager::Heartbeat() {
   auto heartbeat_data = std::make_shared<HeartbeatTableData>();
   SchedulingResources &local_resources = cluster_resource_map_[self_node_id_];
   heartbeat_data->set_client_id(self_node_id_.Binary());
+
   // TODO(atumanov): modify the heartbeat table protocol to use the ResourceSet directly.
   // TODO(atumanov): implement a ResourceSet const_iterator.
-  for (const auto &resource_pair :
-       local_resources.GetAvailableResources().GetResourceMap()) {
-    heartbeat_data->add_resources_available_label(resource_pair.first);
-    heartbeat_data->add_resources_available_capacity(resource_pair.second);
+  if (!(light_heartbeat_enabled_ &&
+        last_heartbeat_resources_.GetAvailableResources().IsEqual(
+            local_resources.GetAvailableResources()))) {
+    for (const auto &resource_pair :
+         local_resources.GetAvailableResources().GetResourceMap()) {
+      heartbeat_data->add_resources_available_label(resource_pair.first);
+      heartbeat_data->add_resources_available_capacity(resource_pair.second);
+    }
+    last_heartbeat_resources_.SetAvailableResources(
+        ResourceSet(local_resources.GetAvailableResources()));
   }
-  for (const auto &resource_pair : local_resources.GetTotalResources().GetResourceMap()) {
-    heartbeat_data->add_resources_total_label(resource_pair.first);
-    heartbeat_data->add_resources_total_capacity(resource_pair.second);
+  if (!(light_heartbeat_enabled_ && last_heartbeat_resources_.GetTotalResources().IsEqual(
+                                        local_resources.GetTotalResources()))) {
+    for (const auto &resource_pair :
+         local_resources.GetTotalResources().GetResourceMap()) {
+      heartbeat_data->add_resources_total_label(resource_pair.first);
+      heartbeat_data->add_resources_total_capacity(resource_pair.second);
+    }
+    last_heartbeat_resources_.SetTotalResources(
+        ResourceSet(local_resources.GetTotalResources()));
   }
 
   local_resources.SetLoadResources(local_queues_.GetResourceLoad());
-  for (const auto &resource_pair : local_resources.GetLoadResources().GetResourceMap()) {
-    heartbeat_data->add_resource_load_label(resource_pair.first);
-    heartbeat_data->add_resource_load_capacity(resource_pair.second);
+  if (!(light_heartbeat_enabled_ && last_heartbeat_resources_.GetLoadResources().IsEqual(
```

Comment:
For clarity, please write

```
if (light_heartbeat_enabled_) {
...
} else {
...
}
```
Instead of mixing the flag logic up with the other conditionals.

Reply:
we may end this with 
```
if (light_heartbeat_enabled) {
  if (changed) {
    // set resources
  }
} else {
  // set resources
}
```
There're two `set resources` code segment which are same. Is it acceptable? @ericl